kernel/src/hil/symmetric_encryption.rs

```diff
@@ -1,42 +1,128 @@
+
+    /// Begin a new message (with the configured IV) when `crypt()` is
+    /// next called.  Multiple calls to `crypt()` may be made between
+    /// calls to `start_message()`, allowing the encryption context to
+    /// extend over non-contiguous extents of data.
+    ///
+    /// If an encryption operation is in progress, this method instead
+    /// has no effect.
+    fn start_message(&self);
+
+    /// Request an encryption/decryption
+    ///
+    /// If the source buffer is not `None`, the encryption input
+    /// will be that entire buffer.  Otherwise the destination buffer
+    /// at indices between `start_index` and `stop_index` will
+    /// provide the input, which will be overwritten.
+    ///
+    /// If `None` is returned, the client's `crypt_done` method will eventually
+    /// be called, and the portion of the data buffer between `start_index`
+    /// and `stop_index` will hold the result of the encryption/decryption.
+    ///
+    /// If `Some(result, source, dest)` is returned, `result` is the
+    /// error condition and `source` and `dest` are the buffers that
+    /// were passed to `crypt`.
+    ///
+    /// The indices `start_index` and `stop_index` must be valid
+    /// offsets in the destination buffer, and the length
+    /// `stop_index - start_index` must be a multiple of
+    /// `AES128_BLOCK_SIZE`.  Otherwise, `Some(EINVAL, ...)` will be
+    /// returned.
+    ///
+    /// If the source buffer is not `None`, its length must be
+    /// `stop_index - start_index`.  Otherwise, `Some(EINVAL, ...)`
+    /// will be returned.
+    ///
+    /// If an encryption operation is already in progress,
+    /// `Some(EBUSY, ...)` will be returned.
+    ///
+    /// For correct operation, the methods `set_key` and `set_iv` must have
+    /// previously been called to set the buffers containing the
+    /// key and the IV (or initial counter value), and a method `set_mode_*()`
+    /// must have been called to set the desired mode.  These settings persist
+    /// across calls to `crypt()`.
+    ///
+    fn crypt(&'a self,
+             source: Option<&'a mut [u8]>,
+             dest: &'a mut [u8],
+             start_index: usize,
+             stop_index: usize)
+             -> Option<(ReturnCode, Option<&'a mut [u8]>, &'a mut [u8])>;
 }
 
-pub trait Client {
-    /// send back to result of the encryption/decryption to the capsule
-    /// this should be hardware independent if the cryptostate is used for all
-    /// implementations
-    fn crypt_done(&self,
-                  data: &'static mut [u8],
-                  dmy: &'static mut [u8],
-                  len: usize)
-                  -> ReturnCode;
+pub trait AES128Ctr {
+    /// Call before `AES128::crypt()` to perform AES128Ctr
+    fn set_mode_aes128ctr(&self, encrypting: bool);
+}
+
+pub trait AES128CBC {
+    /// Call before `AES128::crypt()` to perform AES128CBC
+    fn set_mode_aes128cbc(&self, encrypting: bool);
+}
+
```

Comment:
Sorry, I don't want to be annoying but I would like the see consistency in the naming of the traits.
I think it has been discussed earlier that the first letter should uppercase and the lower?

So, I suggest:
* Aes128Ctr
* Aes128Cbc

Thoughts?